src/crimson/os/seastore/seastore.cc

```diff
@@ -812,32 +833,51 @@ SeaStore::tm_ret SeaStore::_setattrs(
 {
   logger().debug("{} onode={}",
                 __func__, *onode);
-  auto layout = onode->get_mutable_layout(*ctx.transaction);
-  for (auto& [key, val] : aset) {
+  auto& layout = onode->get_mutable_layout(*ctx.transaction);
+  for (auto it = aset.begin(); it != aset.end();) {
+    auto& key = it->first;
+    auto& val = it->second;
+    logger().debug("{} onode={}, key={}",
+                  __func__, *onode, key);
     if (key == OI_ATTR) {
       if (__builtin_expect(
-          val.length() > onode_layout_t::MAX_OI_LENGTH,
-          false)) {
-        logger().error("{} onode={} oi attr too long!");
-        return crimson::ct_error::input_output_error::make();
+          val.length() <= onode_layout_t::MAX_OI_LENGTH,
+          true)) {
+        layout.oi_size = val.length();
+        __builtin_memcpy(&layout.oi[0], val.c_str(), val.length());
+        it = aset.erase(it);
+        continue;
+      } else {
+        layout.oi_size = 0;
       }
-      layout.oi_size = val.length();
-      val.copy_out(0, val.length(), &layout.oi[0]);
     } else if (key == SS_ATTR) {
       if (__builtin_expect(
-          val.length() > onode_layout_t::MAX_SS_LENGTH,
-          false)) {
-        logger().error("{} onode={} oi attr too long!");
-        return crimson::ct_error::input_output_error::make();
+          val.length() <= onode_layout_t::MAX_SS_LENGTH,
+          true)) {
+        layout.ss_size = val.length();
+        __builtin_memcpy(&layout.ss[0], val.c_str(), val.length());
+        it = aset.erase(it);
+        continue;
+      } else {
+        layout.ss_size = 0;
       }
-      layout.ss_size = val.length();
-      val.copy_out(0, val.length(), &layout.ss[0]);
-    } else {
-      //FIXME: right now, only OI_ATTR and SS_ATTR are supported
-      assert(0 == "only OI_ATTR and SS_ATTR are supported for now");
     }
+    it++;
```

Comment:
does it repeat to written to layout.io[0] and layout.ss[0],  I can't figure out the break condition.

Reply:
break condition is "it == aset.end()"

Reply:
so every time overwritten to layout.ai[0] and layout.ss[0]?

Reply:
?
No, only when the key is OI_ATTR or SS_ATTR

Reply:
if in aset there are both keys are OI_ATTR, what happens?  I think the code should avoid this condition.

Reply:
I don't quite follow, aset is an std::map in which there won't be two identical key, right?

Reply:
I just see this function logical. I think we shouldn't assume what input is.  For this function logical, it maybe overwritten layout.oi[0] and layout.ss[0].

Reply:
could you please check if _builtin_xx function has the same behavior in different complier such as clang?  Is it better to use _builtin than libc?

Reply:
I saw some document said : "The compiler determines whether to use the builtin or to call to libc based on the alignment and size of the memory region and compilation parameters such as the target architecture."   So seems no problem here.